### PR TITLE
[KV Connector] Add KDA 4-state support to NixlConnector for Kimi Linear PD disaggregation

### PR DESCRIPTION
**Pull request by JaredforReal**
# Summary

This PR adds KDA (Kimi Delta Attention) 4-state transfer support to the NixlConnector, enabling PD (prefill/decode) disaggregated serving for **Kimi Linear** — a hybrid model that interleaves KDA linear attention layers with MLA layers.

KDA has 4 independent state tensors (`conv_q`, `conv_k`, `conv_v`, `recurrent`) per layer, unlike Mamba2/GDN which use 2 states (fused conv + temporal). The existing NixlConnector only supported 2-state SSM models. This PR generalizes the descriptor building and offset computation to handle KDA's 4-state layout.

# Changes

- **`ssm_conv_transfer_utils.py`**: Add `KDAStateSplitInfo` dataclass and `derive_kda_state_split()` for 4-state offset computation. Dispatch in `derive_mamba_conv_split()` based on `len(mamba_spec.shapes)`. Fix `remote_state_offsets()` to handle `tp_ratio < 0` (P_TP > D_TP).
- **`nixl/worker.py`**: Add `_build_kda_remote()` for KDA hetero-TP descriptor building. Generalize `_build_mamba_local()` to work with variable-length state tuples. Pass raw signed `tp_ratio` to offset computation.
- **`nixl/metadata.py`**: Bump `NIXL_CONNECTOR_VERSION` 4→5. Generalize `ssm_sizes` from `tuple[int, int]` to `tuple[int, ...]`.
- **`utils.py`**: Fix `get_transfer_cache_regions()` to work with variable-length MambaSpec state tuples.
- **`test_nixl_connector_hma.py`**: Add unit tests for `KDAStateSplitInfo` local and remote offsets (positive and negative `tp_ratio`).

# Benchmark Results

Kimi-Linear-48B-A3B-Instruct on 8×H100, ShareGPT dataset, 1000 prompts, request rate = 32 RPS:

| Metric | TP8 baseline | 1P1D (TP4+TP4) | Change |
|--------|-------------|-----------------|--------|
| Mean TTFT | 232.57 ms | 224.71 ms | -3.4% |
| **P99 TTFT** | **587.96 ms** | **337.73 ms** | **-42.5%** |
| **Mean TPOT** | **40.99 ms** | **11.18 ms** | **-72.7%** |
| **P99 TPOT** | **89.80 ms** | **12.34 ms** | **-86.3%** |
| **P99 ITL** | **113.59 ms** | **16.48 ms** | **-85.5%** |
| Output throughput | 5468 tok/s | 5652 tok/s | +3.4% |

PD disaggregation achieves **72-86% reduction in decode latency** (TPOT/ITL) and **42% reduction in P99 TTFT**, with comparable throughput, using the same 8 GPUs.

## Test Plan

- [x] Unit tests for `KDAStateSplitInfo` offset computation (local, tp_ratio>0, tp_ratio<0)
- [x] End-to-end PD disaggregation serving on 8×H100 with `NixlConnector`
- [x] ShareGPT benchmark (1000 prompts) comparing TP8 vs 1P1D TP4

---

# Detailed Serving and Testing:

## 1P1D TP4:

```
# Prefill:
CUDA_VISIBLE_DEVICES=0,1,2,3 \
VLLM_SSM_CONV_STATE_LAYOUT=DS \
VLLM_KV_CACHE_LAYOUT=HND \
VLLM_NIXL_SIDE_CHANNEL_PORT=5560 \
UCX_NET_DEVICES=all \
vllm serve moonshotai/Kimi-Linear-48B-A3B-Instruct \
  --trust-remote-code \
  --port 8100 \
  --host 0.0.0.0 \
  --tensor-parallel-size 4 \
  --gpu-memory-utilization 0.9 \
  --no-disable-hybrid-kv-cache-manager \
  --kv-transfer-config '{"kv_connector":"NixlConnector","kv_role":"kv_both"}' \
  --served-model-name kimi

# Decode
CUDA_VISIBLE_DEVICES=4,5,6,7 \
VLLM_SSM_CONV_STATE_LAYOUT=DS \
VLLM_KV_CACHE_LAYOUT=HND \
VLLM_NIXL_SIDE_CHANNEL_PORT=5660 \
UCX_NET_DEVICES=all \
vllm serve moonshotai/Kimi-Linear-48B-A3B-Instruct \
  --trust-remote-code \
  --port 8200 \
  --host 0.0.0.0 \
  --tensor-parallel-size 4 \
  --gpu-memory-utilization 0.9 \
  --no-disable-hybrid-kv-cache-manager \
  --kv-transfer-config '{"kv_connector":"NixlConnector","kv_role":"kv_both"}' \
  --served-model-name kimi 

# Router
vllm-router --policy round_robin  --vllm-pd-disaggregation   --prefill http://10.1.50.70:8100     --decode http://10.1.50.70:8200     --host 0.0.0.0     --port 8000     --intra-node-data-parallel-size 1
```

## TP8:

```
vllm serve moonshotai/Kimi-Linear-48B-A3B-Instruct/ \
  --tensor-parallel 8 \
  --served-model-name kimi \
  --trust-remote-code \
  --gpu-memory-utilization 0.9 \
  --port 8000 \
  --host 0.0.0.0
```

## 1P1D TP4 ShareGPT request_rate=32 num_prompts=1000

```
vllm bench serve   --backend openai-chat   --base-url http://127.0.0.1:8000   --endpoint /v1/chat/completions   --model kimi   --tokenizer /cloud/oss_checkpoints/moonshotai/Kimi-Linear-48B-A3B-Instruct/   --dataset-name sharegpt   --dataset-path ./data/sharegpt/ShareGPT_V3_unfiltered_cleaned_split.json   --num-prompts 1000   --request-rate 32   --temperature 0   --top-p 1.0   --ignore-eos   --save-result   --result-dir ./bench_results   --result-filename tp8_sharegpt_1000.json   --metadata mode=pd prefill_tp=4 decode_tp=4 --trust-remote-code

============ Serving Benchmark Result ============
Successful requests:                     1000
Failed requests:                         0
Request rate configured (RPS):           32.00
Benchmark duration (s):                  35.28
Total input tokens:                      222179
Total generated tokens:                  199415
Request throughput (req/s):              28.34
Output token throughput (tok/s):         5652.24
Peak output token throughput (tok/s):    8213.00
Peak concurrent requests:                134.00
Total token throughput (tok/s):          11949.70
---------------Time to First Token----------------
Mean TTFT (ms):                          224.71
Median TTFT (ms):                        223.58
P99 TTFT (ms):                           337.73
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          11.18
Median TPOT (ms):                        11.41
P99 TPOT (ms):                           12.34
---------------Inter-token Latency----------------
Mean ITL (ms):                           11.12
Median ITL (ms):                         11.29
P99 ITL (ms):                            16.48
==================================================
```

## TP8 ShareGPT request_rate=32 num_promts=1000

```
vllm bench serve   --backend openai-chat   --base-url http://127.0.0.1:8000   --endpoint /v1/chat/completions   --model kimi   --tokenizer /cloud/oss_checkpoints/moonshotai/Kimi-Linear-48B-A3B-Instruct/   --dataset-name sharegpt   --dataset-path ./data/sharegpt/ShareGPT_V3_unfiltered_cleaned_split.json   --num-prompts 1000   --request-rate 32   --temperature 0   --top-p 1.0   --ignore-eos   --save-result   --result-dir ./bench_results   --result-filename tp8_sharegpt_1000.json   --metadata mode=tp8 tp=8 --trust-remote-code

============ Serving Benchmark Result ============
Successful requests:                     1000
Failed requests:                         0
Request rate configured (RPS):           32.00
Benchmark duration (s):                  36.47
Total input tokens:                      222179
Total generated tokens:                  199415
Request throughput (req/s):              27.42
Output token throughput (tok/s):         5468.42
Peak output token throughput (tok/s):    16121.00
Peak concurrent requests:                332.00
Total token throughput (tok/s):          11561.09
---------------Time to First Token----------------
Mean TTFT (ms):                          232.57
Median TTFT (ms):                        211.69
P99 TTFT (ms):                           587.96
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          40.99
Median TPOT (ms):                        37.39
P99 TPOT (ms):                           89.80
---------------Inter-token Latency----------------
Mean ITL (ms):                           33.94
Median ITL (ms):                         15.93
P99 ITL (ms):                            113.59
==================================================
```

### CODE REVIEW

**Comment by ZhanqiuHu** [comment]
Thanks for the PR! Would it be possible to add Linear as a new mamba type (something like this) and reduce the changes? 
```
if mamba_spec.mamba_type == MambaAttentionBackendEnum.LINEAR:
        # KDA: 4 shapes = (conv_q, conv_k, conv_v, recurrent).
        # First 3 are 2D conv states in DS layout, same conv_rows/dtype.
        # Maps directly to 3 sub-projections + SSM temporal.
        assert len(mamba_spec.shapes) == 4, (
            f"KDA/Linear expects 4 state tensors, got {len(mamba_spec.shapes)}"
        )
        assert is_conv_state_dim_first(), "KDA transfer requires DS layout"
        conv_rows = mamba_spec.shapes[0][1]
        local_proj_dims = (
            mamba_spec.shapes[0][0],  # q_dim_local
            mamba_spec.shapes[1][0],  # k_dim_local
            mamba_spec.shapes[2][0],  # v_dim_local
        )
        conv_dtype_size = torch.tensor([], dtype=mamba_spec.dtypes[0]).element_size()
        conv_state_bytes = sum(
            torch.Size(s).numel() * torch.tensor([], dtype=d).element_size()
            for s, d in zip(mamba_spec.shapes[:3], mamba_spec.dtypes[:3])
        )
        ssm_state_bytes = (
            torch.Size(mamba_spec.shapes[3]).numel()
            * torch.tensor([], dtype=mamba_spec.dtypes[3]).element_size()
        )
        return MambaConvSplitInfo(
            conv_rows=conv_rows,
            local_proj_dims=local_proj_dims,
            conv_dtype_size=conv_dtype_size,
            ssm_sizes=(conv_state_bytes, ssm_state_bytes),
        )
```

**Review by NickLucche — COMMENTED** [review]
Thanks for the PR @JaredforReal , this is already in a great shape!
I think we may want to add it to
```
    _supported = (
        MambaAttentionBackendEnum.MAMBA2,
        MambaAttentionBackendEnum.GDN_ATTN,
    )
```
and handle like @ZhanqiuHu to keep logic consistent

**Comment by NickLucche** [comment]
PS @JaredforReal did you test with an heterogeneous setup? Getting some weird answers on the first run

**Comment by JaredforReal** [comment]
@NickLucche Sorry, the PR description of tp_ratio is really confusing. I have implemented a unit test for tp_ratio>0/<0, but I didn't run a heterogeneous setup E2E test. It turns out that the block size is incompatible with TP=2(480) and TP=4(944), causing the weird output
@ZhanqiuHu Thanks for the suggestion
I'm working on it

**Comment by JaredforReal** [comment]
@ZhanqiuHu KDA is now refactored into KimiGatedDeltaNetAttention(GatedDeltaNetAttention), witch make its `MambaAttentionBackendEnum` to be `GDN_ATTN`.
I can delete `KDAStateSplitInfo` and reuse `MambaConvSplitInfo`, but it still needs `if len(mamba_spec.shapes) == 4` to deal with KDA's special state shape

**Comment by JaredforReal** [comment]
@NickLucche, I’d like to implement a heterogeneous TP setup in a follow-up PR. I’ve been busy with other events recently, but I’ll take this up next.

**Comment by JaredforReal** [comment]
@ZhanqiuHu @NickLucche  PTAL, thanks!

**Comment by ZhanqiuHu** [comment]
Thanks for the prompt updates! Is it possible to keep the original 2 tuple ssm sizes since conv offsets already encodes the 3 component in conv. Then we don't need to change the metadata.py and worker.py.

File: vllm/distributed/kv_transfer/kv_connector/v1/ssm_conv_transfer_utils.py [hunk 1]
```diff
@@ -129,12 +130,52 @@ def derive_mamba_conv_split(
...
+            * torch.tensor(
+                [],
+                dtype=mamba_spec.dtypes[3],  # type: ignore[misc]
+            ).element_size()
+        )
+        return MambaConvSplitInfo(
+            conv_rows=conv_rows,
+            local_proj_dims=local_proj_dims,
+            conv_dtype_size=conv_dtype_size,
+            ssm_sizes=(conv_state_bytes, ssm_state_bytes),
```
Comment (NickLucche) [inline vllm/distributed/kv_transfer/kv_connector/v1/ssm_conv_transfer_utils.py]: ok this can be a 2 element tuple again then

**Review by NickLucche — COMMENTED** [review]
>@NickLucche, I’d like to implement a heterogeneous TP setup in a follow-up PR. I’ve been busy with other events recently, but I’ll take this up next.

@JaredforReal sg! Could you add a quick check in `_validate_remote_agent_handshake` checking whether KDA and tp_ratio != 1 (heteroTP) is used, to raise as NotSupported for now?

Also, I see we can go back to using a 2-element tuple with latest changes.
Other than that this is LGTM

File: vllm/distributed/kv_transfer/kv_connector/v1/ssm_conv_transfer_utils.py (hunk 1 above)
Comment (JaredforReal) [reply vllm/distributed/kv_transfer/kv_connector/v1/ssm_conv_transfer_utils.py]: will work on it, thanks!

**Comment by JaredforReal** [comment]
The KDA PD disaggregation support has been done by fusing KDA's conv QKV into one element to align with GDN format. See #44539 
This PR is no longer needed
